### ISSUE TITLE
[ENH] technical roadmap 2025

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#7707]
Umbrella issue for collecting, consolidating, and prioritizing roadmap items for the 2025 increment.

How to contribute:

* user or developer - community suggestions appreciated in this thread!
* new to open source and want to contribute code? Check if you see sth interesting and get in touch on [discord](https://discord.com/invite/54ACzaFsn7) - at meet-ups, workstream meetings, or just chat

**Comment by fkiraly** [comment]
From developer roadmap planning January 24

(to be used to prioritize)

- [name=Felipe Angelim - core dev]
- [name=Revanth - core dev]
- [name=Franz Kiraly - core dev & council]
- [name=Satvik - mentee]
- [name=Aryan - mentee] 
- [name=M. Armaghan Shakir - core dev]
- [name=Vedant Agrawal - mentee]
- [name=Jigyasu - mentee]
- [name=Andrea Gobbi - co-mentor, dsipts dev]
- [name=Anirban Ray - core dev]
- [name=Marc Rovira - core dev & council]
- [name=Ugo Onyeka - council]
- [name=Sandeep - dsipts dev]
- [name=Julian Fong - dev]
- [name=Simon Blanke - dev hyperactive]
- [name=Pranav Bhat - mentee]
- [name=Mridul Jain]

:+1: = subjective prio (3 per participant)
:heavy_check_mark: = working on this (current plan)



**Foundation models** (12) :+1::+1::+1::+1::+1::+1::+1::heavy_check_mark:

* generalize HFForecaster / HFAdapter :+1::+1::+1::heavy_check_mark:
    * current `HFTransformerForecaster` only works for Autoformer, Informer, TimeSeriesTransformer
    * bunch of other huggingface/transformer based estimators can be supported like: TimeLLM, Time-Moe, PatchTST, TinyTimeMixer, TimesFM, PatchTSMixer, Time-Mixer
    * improve the current design for `fit_strategy` methods

* PeftAdapter :+1::+1::heavy_check_mark:
    * wrapper around `HFTransformerForecaster` and other models that can support peft fine-tuning
    * currenly `HFTransformerForecaster` takes `fit_strategy="peft"`

* foundation models full design, rebase current Fm on this


**Deep Learning, torch, ptf, dsipts** (16) :+1::+1::+1::+1::+1::+1::heavy_check_mark::heavy_check_mark:

* interfacing Deep-Learning based models:+1::+1::+1::+1:
    * add new models to sktime and pytorch-forecasting :heavy_check_mark: 
    * draw line between keeping the models in sktime and pytorch-forecasting
    * also enable integration betweent the two libraries

* pytorch-forecasting/sktime integration:+1:
    * unified data containers for torch & forecasting
    * rework with dsipts and TSLib (TSLib has some architectures not present in dsipts)
    * rebasing sktime models on the new architecture

* pytorch-forecasting/sktime integration (DSIP SIDE)::+1: :+1::+1::+1:
    * embark some *nice to have* features such as:
        * parallel training session (same model, different parameters, same dataset) or (same dataset,m different models)
        * optuna for newtork parmaeter optimization
        * comparison between models
        * consider *hydra* --> sweepers for optuna and for slurm based cluster GPU based
        * stacked generalization models: use outputs of N models for traing the final model (some architectures are good at finding spikes some others for long trend)
        * loss functions for timeseries, optimizer for timeseries
        * having model and trainer params directly accessible through estimators' `get_params` and `set_params` interface. (is it possible to have dictionaries behave as 'dummy estimators'?)

* A suggestion : Refactor various `PyTorchDataset` in `sktime`(issue: https://github.com/sktime/sktime/issues/7598)

:+1::+1::+1: **Forecasting base framework** (9)

* improve support of Darts provided models (more interfaces, support for past/future/static covariates) :+1: 
    * different type of X can be a generic feature, neuralforecast also use that

* rework/improve `ForecastingHorizon`, decouple from `pandas` :+1: :+1::+1::+1:

* improve logging in the framework (almost no logs come from framework, very difficult to debug on production) :+1:

**Reduction models**:+1::+1::+1:(3)

* fix all the problems with the reducers :sweat_smile:
    * Related to current "global forecasting" terminology conflict 

* resolve API design for global reduction models! Currently subclass, could be merged or redesigned.

:+1: :+1: :+1: **Hierarchical models** (3)

* hierarchical reconciliation API - adding new reconciliation strategies and revisiting the current `transform` interface :heavy_check_mark:

* skpro: distributions for hierarchical index, broadcasting, concat, etc - required for forecasts that are both distributional and probabilistic
* Adding temporal hierarchical reconciliation capabilities to sktime 

:+1: :+1: :+1:**Benchmarking, Tuning** (4)

* finish benchmarking rework. (where did Hazrul go?)

* Adding code to reproduce and benchmark models with M1-M5 competition data.
    * Currently, some estimators in sktime are not able to reproduce experiments where the maximum time index is not equal to all panels. This is caused by having a "global" cutoff parameter, that is used to parse relative forecast horizons (verified this with reducers)

* abstract tuning engine - have optuna tuning for all learning tasks. with `hyperactive` package :+1::heavy_check_mark:


**Maintenance and release management**:+1: :+1: (2)

* improve release strategy (pre-planned feature targeted minor releases, feature to go to minor instead of patch, plan major release, etc.) 
* consideration of reducing package size by splitting by module, especially that now few forks are there 
* smaller item: merging sktime pykalman fork back into pykalman

**Comment by dhruvd-1** [comment]
hi can this pls be assigned to me

**Comment by fkiraly** [comment]
@dhruvd-1, this is the entire 2025 roadmap, this is not for assigning. This is an umbrella issue.

As the description says, this is a "good first issue" in the sense that you can pick something from it if interested. You can ask on discord and we can suggest a starter item.
